src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java

```diff
@@ -458,26 +465,29 @@ private void loadStagingTable(TableId stagingTableId, TableBlob blob, int attemp
   }
 
   private void mergeStagingTable(TableId stagingTableId, TableBlob blob,
-                                 int attemptNumber) throws InterruptedException {
+                                 int attemptNumber) throws InterruptedException, DeltaFailureException {
     LOG.debug("Merging batch {} for {}.{}", blob.getBatchId(), blob.getDataset(), blob.getTable());
     TableId targetTableId = TableId.of(project, blob.getDataset(), blob.getTable());
-    Table targetTable = bigQuery.getTable(targetTableId);
-    if (targetTable == null) {
-      // We require that the table already exists and that it was created by this plugin when
-      // handling a DDL create table event.
-      // this is because the primary key for the table is contained in the DDL event, but not in DML events.
-      // so the primary key info is stored in the table definition.
-      // If we had the primary key information here, we could create the table now if it didn't exist for some reason
-      throw new IllegalStateException(
-        String.format("No target table %s.%s found. Please ensure that tables created by the replicator are not "
-                        + "deleted by other processes.", blob.getDataset(), blob.getTable()));
-    }
+    List<String> primaryKey = primaryKeyStore.get(targetTableId);
+    if (primaryKey == null) {
+      byte[] pkBytes;
+      try {
+        pkBytes = context.getState(String.format("%s-%s-primary-key", targetTableId.getDataset(),
+                                                 targetTableId.getTable()));
+      } catch (Exception e) {
+        throw new DeltaFailureException(
```

Comment:
DeltaFailureException will cause the replicator to fail immediately, and should only be thrown if there is no way for the replicator to ever make progress in any scenario. 

For this case, it could potentially be a temporary issue, so we can throw the exception as-is, and let the replicator handle retrying.

Reply:
Sure, makes sense. will do.